src/Tasks/SystemState.cs

```diff
@@ -573,6 +592,38 @@ private string[] GetDirectories(string path, string pattern)
             return getDirectories(path, pattern);
         }
 
+        /// <summary>
+        /// Cached implementation of GetFiles aimed to verify existence of a file in a directory.
+        /// It does not throw if directory do not exists.
+        /// </summary>
+        /// <param name="path"></param>
+        /// <param name="fileName"></param>
+        /// <returns>file full path or null if file do not exists</returns>
+        private string GetDirectoryFile(string path, string fileName)
+        {
+
+            instanceLocalDirectoryFiles.TryGetValue(path, out Dictionary<string, string> cached);
+            if (cached == null)
+            {
+                string[] files;
+                try
+                {
+                    files = getFiles(path, "*");
+                }
+                catch(DirectoryNotFoundException)
+                {
+                    files = Array.Empty<string>();
+                }
+
+                cached = files.ToDictionary(fn => Path.GetFileName(fn), StringComparer.OrdinalIgnoreCase);
+                instanceLocalDirectoryFiles[path] = cached;
+            }
+
+            cached.TryGetValue(fileName, out string fullPathFileName);
```

Comment:
If it fails, should we not find it and add it?

Reply:
This does exactly what we want, return file full path if file is present or null if it does not exists.
On the other hand, it is unusual usage pattern for `TryGetValue` - I will create comment here to describe it.